src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs

```diff
@@ -6322,12 +6561,190 @@ private IOperation VisitNoneOperationExpression(IOperation operation)
+                // else goto next block;
+                LeaveRegionsUpTo(resultRegion);
+                AppendNewBlock(new BasicBlockBuilder(BasicBlockKind.Block), linkToPrevious: true);
+            }
+            else
+            {
+                LeaveRegionsUpTo(resultRegion);
+            }
+
+            if (operation.HandlerCreationHasSuccessParameter)
+            {
+                RemovePlaceholder(outParameterPlaceholder!);
+            }
+
+            var appendCalls = collectAppendCalls(operation);
+
+            int appendCallsLength = appendCalls.Length;
+            for (var i = 0; i < appendCallsLength; i++)
+            {
+                var appendRegion = new RegionBuilder(ControlFlowRegionKind.LocalLifetime);
+                EnterRegion(appendRegion);
+                var appendCall = appendCalls[i];
+                if (operation.HandlerAppendCallsReturnBool)
+                {
+                    Debug.Assert(resultBlock != null);
+
+                    if (i == appendCallsLength - 1)
+                    {
+                        // No matter the result, we're going to the result block next. So just visit the statement, and if the current block can be
+                        // combined with the result block, the compaction machinery will take care of it
+                        VisitStatement(appendCall.AppendCall);
+                        AppendNewBlock(resultBlock, linkToPrevious: true);
+                    }
+                    else
+                    {
+                        // if (!appendCall()) goto result else goto next block
+                        ConditionalBranch(VisitRequired(appendCall.AppendCall), jumpIfTrue: false, resultBlock);
+                        AppendNewBlock(new BasicBlockBuilder(BasicBlockKind.Block), linkToPrevious: true);
+                    }
+                }
+                else
+                {
+                    VisitStatement(appendCall.AppendCall);
+
+                    if (i == appendCallsLength - 1 && operation.HandlerCreationHasSuccessParameter)
+                    {
+                        Debug.Assert(resultBlock != null);
+                        AppendNewBlock(resultBlock, linkToPrevious: true);
+                    }
+                }
+                LeaveRegionsUpTo(appendRegion);
+                LeaveRegion();
+            }
+
+            if (appendCallsLength == 0 && operation.HandlerCreationHasSuccessParameter)
+            {
+                Debug.Assert(resultBlock != null);
+                AppendNewBlock(resultBlock, linkToPrevious: true);
+            }
+
+            _currentInterpolatedStringHandlerId = previousHandlerId;
+            return new FlowCaptureReferenceOperation(handlerCaptureId, operation.Syntax, operation.Type, operation.GetConstantValue());
+
+            static ImmutableArray<IInterpolatedStringAppendOperation> collectAppendCalls(IInterpolatedStringHandlerCreationOperation creation)
```

Comment:
>ImmutableArray<IInterpolatedStringAppendOperation> [](http://example.com/codeflow?start=19&length=50)

Why does this have to be an immutable array. The value is used locally and it looks like leaving it as a builder would be just fine.

Reply:
Personal preference of not having to then manage the lifetime of the builder across functions.

Reply:
> Personal preference of not having to then manage the lifetime of the builder across functions.

I do not think the gained simplicity outweighs the fact that we unnecessary allocate an array. Creation of the builder can be pulled to the call site to make lifetime management contract clear.